src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,331 @@
+            res = kwargs.pop("res", None)
+
+            if compressor and extents and res:
+                compressor_args = dict(compressor)
+                compressor_name = compressor_args.pop("_name")
+                if compressor_name:
+                    compressor_args = {
+                        k: None if not v else v for k, v in compressor_args.items()
+                    }
+                    compression_filter = vars(filter).get(compressor_name)(
+                        **compressor_args
+                    )
+                    compression_level = compression_filter.level
+
+                else:
+                    raise ValueError
+
+                # create empty array, inputs is a single reference file when creating the array
+                with rasterio.open(inputs) as src:
+                    dst_profile = src.profile.copy()
+                    dst_profile.update({
+                        "driver": "TileDB",
+                        "transform": from_origin(extents.minx, extents.maxy, *res),
+                        "width": math.floor((extents.maxx - extents.minx) / res[0]),
+                        "height": math.floor((extents.maxy - extents.miny) / res[1]),
+                        "blockysize": tile_size,
+                        "blockxsize": tile_size,
+                    })
+                    with rasterio.open(output, "w",
+                                       COMPRESSION=compressor_name[:-6].upper(), # remove trailing `Filter`
+                                       COMPRESSION_LEVEL=compression_level,
+                                       **dst_profile,):
+                        pass
+        else:
+            # srcs and dst have same number of bands
+            resampling = rasterio.enums.Resampling[kwargs.pop("resampling", "bilinear").lower()]
+            with rasterio.open(output, mode="r+") as dst:
+                for b in range(dst.count):
+                    for blk, srcs in inputs:
+                        input_datasets = []
+                        try:
+                            input_datasets.extend([rasterio.open(s) for s in srcs])
+                            row_off = blk[0][0]
+                            col_off = blk[1][0]
+                            dst_window = rasterio.windows.Window(col_off, row_off,
+                                                                blk[1][1] - col_off, blk[0][1] - row_off)
+                            dst_bounds = rasterio.windows.bounds(dst_window, dst.transform)
+
+                            if len(srcs) == 1:
+                                with rasterio.vrt.WarpedVRT(input_datasets[0], crs=dst.crs, transform=dst.transform, resampling=resampling, width=dst.width, height=dst.height, masked=True) as vrt:
+                                    data = vrt.read(window=dst_window)
+                            else:
+                                data, _ = rasterio.merge.merge(input_datasets, bounds=dst_bounds, res=dst.res,
```

Comment:
@normanb I think it would be better to use the same merge tool regardless of the number of inputs. Rasterio isn't currently testing that WarpedVRT and merge.merge produce the exact same results.

Does merge.merge work well here? I didn't have TileDB in mind, but I think it ought to be supported well enough.

Reply:
I couldn't get it to work with a single input file which is why I used `WarpedVRT`, the return from a single input to `merge.merge` is an array of zeros.

Reply:
That feels like a rasterio bug.

Speaking of array of zeros, this line here https://github.com/rasterio/rasterio/blob/main/rasterio/merge.py#L293 means that the "nodata" value of an output array isn't properly configurable like file output is. That's something I should fix, too. Is that what you're experiencing @normanb? That passing `dst.nodata` to `merge()` has no effect?

Reply:
For 1.4.0, I propose to do this in Rasterio: https://github.com/rasterio/rasterio/issues/1867. That would be useful for TileDB raster ingest if "r+" mode was also supported.